thumberd/thumberd.go

```diff
@@ -481,11 +481,31 @@ func thumbServer(w http.ResponseWriter, r *http.Request, sem chan int) {
 	}
 	defer srcReader.Body.Close()
 
+	fmt.Printf("%#v\n", params)
+
+	w.Header().Set("Last-Modified", time.Now().UTC().Format(http.TimeFormat))
+
+	imageBlob, format, err := fetchImageWithCorrectFormat(srcReader.Body)
+	if err != nil {
+		message := "Fetch image failed: " + err.Error()
+		glog.Error(message, http.StatusInternalServerError)
+		http.Error(w, message, http.StatusInternalServerError)
+		atomic.AddInt64(&http_stats.thumb_error, 1)
+		return
+	}
+
 	content_type := ""
 	switch params.FormatOutput {
 	case "":
-		content_type = srcReader.Header.Get("Content-Type")
-		content_type, _ = url.QueryUnescape(content_type)
+		if supported_content_type, ok := getContentTypeFromFormat()[format]; ok {
+			content_type = supported_content_type
+		} else {
+			message := "the image format not supported"
+			glog.Error(message)
+			http.Error(w, message, http.StatusInternalServerError)
+			atomic.AddInt64(&http_stats.thumb_error, 1)
+			return
+		}
```

Comment:
If request doesn't assign target format, use the detected image content type instead of Content-Type from request header.

So it will make sure response Content-Type is image MIME-TYPE, exclude XSS possibility.